// @nexus/schema/dist/extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/**
 * Container object living on `fieldDefinition.extensions.nexus`
 */
class NexusFieldExtension {
    constructor(config) {
        const { resolve } = config, rest = tslib_1.__rest(config, ["resolve"]);
        this.config = rest;
        this.hasDefinedResolver = Boolean(resolve);
    }
}
exports.NexusFieldExtension = NexusFieldExtension;
/**
 * Container object living on `inputObjectType.extensions.nexus`
 */
class NexusInputObjectTypeExtension {
    constructor(config) {
        const { definition } = config, rest = tslib_1.__rest(config, ["definition"]);
        this.config = rest;
    }
}
exports.NexusInputObjectTypeExtension = NexusInputObjectTypeExtension;
/**
 * Container object living on `objectType.extensions.nexus`
 */
class NexusObjectTypeExtension {
    constructor(config) {
        const { definition } = config, rest = tslib_1.__rest(config, ["definition"]);
        this.config = rest;
    }
}
exports.NexusObjectTypeExtension = NexusObjectTypeExtension;
/**
 * Container object living on `interfaceType.extensions.nexus`
 */
class NexusInterfaceTypeExtension {
    constructor(config) {
        const { definition } = config, rest = tslib_1.__rest(config, ["definition"]);
        this.config = rest;
    }
}
exports.NexusInterfaceTypeExtension = NexusInterfaceTypeExtension;
/**
 * Container object living on `schema.extensions.nexus`. Keeps track
 * of metadata from the builder so we can use it when we
 */
class NexusSchemaExtension {
    constructor(config) {
        this.config = config;
    }
}
exports.NexusSchemaExtension = NexusSchemaExtension;
